Penumbra/UI/SharedTagManager.cs

```diff
@@ -0,0 +1,160 @@
+            false, true) || _isPopupOpen)
+            return DrawSharedTagsPopup(localTags, modTags, editLocal);
+
+
+        return tagToAdd;
+    }
+
+    private string DrawSharedTagsPopup(IReadOnlyCollection<string> localTags, IReadOnlyCollection<string> modTags, bool editLocal)
+    {
+        var selected = "";
+        if (!ImGui.IsPopupOpen(PopupContext))
+        {
+            ImGui.OpenPopup(PopupContext);
+            _isPopupOpen = true;
+        }
+
+        var display = ImGui.GetIO().DisplaySize;
+        var height = Math.Min(display.Y / 4, 10 * ImGui.GetFrameHeightWithSpacing());
+        var width = display.X / 6;
+        var size = new Vector2(width, height);
+        ImGui.SetNextWindowSize(size);
+        using var popup = ImRaii.Popup(PopupContext);
+        if (!popup)
+            return selected;
+
+        ImGui.Text("Shared Tags");
+        ImGuiUtil.HoverTooltip("Right-click to close popup");
+        ImGui.Separator();
+
+        foreach (var tag in SharedTags)
+        {
+            if (DrawColoredButton(localTags, modTags, tag, editLocal))
+            {
+                selected = tag;
+                return selected;
+            }
+            ImGui.SameLine();
+        }
+
+        if (ImGui.IsMouseClicked(ImGuiMouseButton.Right))
+        {
+            _isPopupOpen = false;
+        }
+
+        return selected;
+    }
+
+    private static bool DrawColoredButton(IReadOnlyCollection<string> localTags, IReadOnlyCollection<string> modTags, string buttonLabel, bool editLocal)
+    {
+        var isLocalTagPresent = localTags.Contains(buttonLabel);
+        var isModTagPresent = modTags.Contains(buttonLabel);
+
+        var buttonWidth = CalcTextButtonWidth(buttonLabel);
+        // Would prefer to be able to fit at least 2 buttons per line so the popup doesn't look sparse with lots of long tags. Thus long tags will be trimmed.
+        var maxButtonWidth = (ImGui.GetContentRegionMax().X - ImGui.GetWindowContentRegionMin().X) * 0.5f - ImGui.GetStyle().ItemSpacing.X;
+        var displayedLabel = buttonLabel;
+        if (buttonWidth >= maxButtonWidth)
+        {
+            displayedLabel = TrimButtonTextToWidth(buttonLabel, maxButtonWidth);
+            buttonWidth = CalcTextButtonWidth(displayedLabel);
+        }
+
+        // Prevent adding a new tag past the right edge of the popup
+        if (buttonWidth + ImGui.GetStyle().ItemSpacing.X >= ImGui.GetContentRegionAvail().X)
+            ImGui.NewLine();
+
+        // Trimmed tag names can collide, but the full tags are guaranteed distinct so use the full tag as the ID to avoid an ImGui moment.
+        ImRaii.PushId(buttonLabel);
+
+        if (editLocal && isModTagPresent || !editLocal && isLocalTagPresent)
+        {
+            using var alpha = ImRaii.PushStyle(ImGuiStyleVar.Alpha, 0.5f);
+            ImGui.Button(displayedLabel);
+            alpha.Pop();
```

Comment:
Again, recommend against early returns when drawing.
You also don't have to call Pop() here because the using/dispose will do that when leaving scope.